docs/events/2023/docker-workshop.md

```diff
@@ -0,0 +1,241 @@
+FROM base as runner
+
+COPY --from=build /app/dist ./dist
+COPY --from=build /app/node_modules ./node_modules
+
+CMD ["node", "dist/app.js"]
+```
+
+This code is the same as the earlier example but uses a multi-stage build to bundle and transpile the code before running it. 
+
+This means that only the specific files copied from the build stage are included in the final image. This means that the final image is much smaller than it would be if you included the build tools in the final image.
+
+## `docker-compose`
+
+Docker compose is a tool that allows you to run multiple docker containers at the same time. This is useful for things like running a database and a web server that need to talk to one another at the same time.
+
+```yaml
+services:
+  tsnode:
+    image: ghcr.io/progsoc/docker-ts-2023:latest
+    ports:
+      - 3000:3000
+```
+
+This is an example of a docker compose file that runs the docker image that we built earlier. It exposes port 3000 on the host machine and runs the docker image.
+
+You might be able to see that the image url is similar to the way that `git` urls work. This is because docker images are stored in a similar way to git repositories. The `:latest` part tells docker to get the latest version of the image that has been pushed to the repository.
+
+You can change it to a specific version if you want to use a specific version of the image e.g. postgres 14 or 15.
+
+## `docker-compose` with multiple services
+
+```yaml
+services:
+  tsnode:
+    image: ghcr.io/progsoc/docker-ts-2023:latest
+    ports:
+      - 3000:3000
+  postgres:
+    image: postgres:14
+    ports:
+      - 5432:5432
+    environment:
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: postgres
+      POSTGRES_DB: postgres
+    volumes:
+      - postgres-data:/var/lib/postgresql/data
+
+volumes:
+    postgres-data:
+```
+
+This is an example of a docker compose file that runs two docker images. The first is the image that we built earlier and the second is a postgres database. A lot of these common infrastructure services have docker images that you can use to run them.
+
+This makes it really easy to define a set of services and run them without any additional steps.
+
+One thing to note is that the `volumes` section at the bottom defines a volume that is used by the postgres image. This is a way of storing data that is outside of the docker image. This is useful for things like databases that need to store data outside of the docker image.
+
+Normally when you run a docker container the data doesn't stay after you restart or stop it, unless you use a volume. This is because the data is stored inside the docker image and when you restart or stop the container the data is lost.
+
+## Building a docker image
+
+To build a docker image you need to run the `docker build` command. This command takes a few different arguments but the most important one is the `.` at the end. This tells docker to build the image from the current directory.
```

Comment:
Probably good to elaborate that the `.` means that it's looking for a file called "Dockerfile" in the folder `.` (the current folder). However, any other folder can be specified too. Also mention what -t is because it's pretty important.